### PR TITLE
Move dependent-type-finding operations out of process.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/43622.  That PR should be reviewed first.

This moves the searching portion of the dependent-type-finder OOP allowing s to do that expensive work without impacting the main VS process.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski This is ready for review now.

File: src/EditorFeatures/Test/SymbolFinder/DependentTypeFinderTests.cs [hunk 1]
```diff
@@ -4,19 +4,58 @@
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test/SymbolFinder/DependentTypeFinderTests.cs:4]: renaming the file made this show as two different files, so i'm going to rename later.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs [hunk 2]
```diff
@@ -174,7 +174,7 @@ internal static partial class DependentTypeFinder
             // We might miss a derived type in C if there's an intermediate derived type
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs:174]: the changes in this file are followup to PR feedback from previous PR.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_ImplementingTypes.cs [hunk 3]
```diff
@@ -67,9 +84,14 @@ static bool TypeMatches(INamedTypeSymbol type, SymbolSet set)
...
-                // FindDerivedInterfacesAsync.
-                return allTypes.WhereAsArray(t => t.TypeKind == TypeKind.Class || t.TypeKind == TypeKind.Struct);
+                // Only classes/struct/delegates/enums implement interface types.  Derived interfaces can be found with
+                // FindDerivedInterfacesAsync.  Delegates/Enums only happen in a few corner cases.  For example, enums
+                // implement IComparable, and delegates implement ICloneable.
+                return allTypes.WhereAsArray(
+                    t => t.TypeKind == TypeKind.Class ||
+                         t.TypeKind == TypeKind.Struct ||
+                         t.TypeKind == TypeKind.Delegate ||
+                         t.TypeKind == TypeKind.Enum);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_ImplementingTypes.cs:94]: found by @jasonmalinowski  when reviewing the the previous PR.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs [hunk 4]
```diff
@@ -484,10 +484,14 @@ internal enum FunctionId
         Renamer_FindRenameLocationsAsync = 387,
         Renamer_ResolveConflictsAsync = 388,
 
-        ChangeSignature_Data = 389,
+        ChangeSignature_Data = 400,
```
Comment (tmat) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs:487]: >400 [](start = 31, length = 3)

Shouldn't this be immutable once shipped? This is used for telemetry, so changing the values would render counters that span multiple version useless.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/IRemoteDependentTypeFinder.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
```
Comment (tmat) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/IRemoteDependentTypeFinder.cs:5]: >using [](start = 0, length = 5)

#nullable enable

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DependentTypeFinder.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
```
Comment (tmat) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DependentTypeFinder.cs:5]: >System [](start = 6, length = 6)

#nullable enable

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_Remote.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Comment (tmat) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_Remote.cs:3]: >// [](start = 0, length = 2)

#nullable enable

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs:487]: OK, as long as we haven't shipped any of these.

---
In reply to: [418199711](https://github.com/dotnet/roslyn/pull/43636#discussion_r418199711) [](ancestors = 418199711)

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/IRemoteDependentTypeFinder.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/IRemoteDependentTypeFinder.cs:5]: will add in: https://github.com/dotnet/roslyn/pull/43804

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DependentTypeFinder.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DependentTypeFinder.cs:5]: will add in https://github.com/dotnet/roslyn/pull/43804

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_Remote.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_Remote.cs:3]: will add in https://github.com/dotnet/roslyn/pull/43804

**Pull request merged**